### PR TITLE
Gps dma dev

### PR DESCRIPTION
**Pull request by filip6464**
Added GPS class to retrieve information about location, time and status of satellite connection.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -15,7 +15,7 @@ set(CMAKE_OBJCOPY arm-none-eabi-objcopy)
 set(CMAKE_OBJDUMP arm-none-eabi-objdump)
 set(SIZE arm-none-eabi-size)
 
-SET(LINKER_SCRIPT ${CMAKE_SOURCE_DIR}/STM32F446RETX_FLASH.ld)
+SET(LINKER_SCRIPT ${CMAKE_SOURCE_DIR}/STM32F446RETX_FLASH.ld Lib/Gps/gps.cpp Lib/Gps/Inc/Gps/gps.hpp)
```
Comment (wiktorwieclaw) [inline CMakeLists.txt:18]: added files need to be removed. When adding files to a project in clion a popup comes up, make sure to uncheck automatic additions to cmake.

File: CMakeLists_template.txt [hunk 2]
```diff
@@ -71,6 +71,7 @@ include_directories(
         Drivers/STM32F4xx_HAL_Driver/Inc
         Middlewares/Third_Party/BME280
         Middlewares/Third_Party/FreeRTOS/Inc
+        Middlewares/Third_Party/UARTDMAUtils
```
Comment (wiktorwieclaw) [inline CMakeLists_template.txt:74]: remove this library if it is not used

File: Inc/obc.hpp [hunk 3]
```diff
@@ -5,13 +5,15 @@
...
 
+
 #include "Buses/i2c.hpp"
 #include "Buses/spi.hpp"
 #include "Buses/uart.hpp"
 #include "Lcd/lcd.hpp"
 #include "MCU_Board/button.hpp"
 #include "MCU_Board/led.hpp"
 #include "Rtc/rtc.hpp"
+#include <Gps/gps.hpp>
```
Comment (wiktorwieclaw) [inline Inc/obc.hpp:16]: "" instead of <>

File: Src/interrupt_handlers.cpp [hunk 4]
```diff
@@ -24,3 +24,14 @@ void USER_GPIO_EXTI_IRQHandler(uint16_t GPIO_Pin) {
         }
     }
 }
+
+void GPS_IRQHandler(UART_HandleTypeDef *huart){
+    if(RESET != __HAL_UART_GET_FLAG(huart, UART_FLAG_IDLE))   //Judging whether it is idle interruption
+    {
+        __HAL_UART_CLEAR_IDLEFLAG(huart);                     //Clear idle interrupt sign (otherwise it will continue to enter interrupt)
```
Comment (wiktorwieclaw) [inline Src/interrupt_handlers.cpp:31]: if possible consider writing something like
```if (__HAL_UART_GET_FLAG(huart, UART_FLAG_IDLE) == SET)```
i think that it is clearer

unnecessary comments. You could leave something like ```// clear to break from interrupt```
keep comments as short as possible.

File: Src/stm32f4xx_it.c [hunk 5]
```diff
@@ -273,7 +272,8 @@ void TIM6_DAC_IRQHandler(void)
 void DMA2_Stream2_IRQHandler(void)
 {
   /* USER CODE BEGIN DMA2_Stream2_IRQn 0 */
-
+  //UARTDMA_DmaIrqHandler(&huartdma);
```
Comment (wiktorwieclaw) [inline Src/stm32f4xx_it.c:275]: don't leave dead code, remember to uncomment or remove this later

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    // Date
+    uint8_t Day;
+    uint8_t Month;
+    uint8_t Year;
+
+    /*ignored because no information in documentation*/
+    // double  MagneticVariation;       //degrees E=east or W=west, East/West Indicator redundand
+
+    char Mode;          // Mode  A= Autonomus D=DGPS, E=DR
+    char* checksum[5];  // Checksum of all chars between $ (start body message) and * (end body
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:44]: std::array

if char* means zero terminated string then use
use std::array<gsl::zstring, 5>

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+#ifndef HABSAT_OBC_GPS_H
+#define HABSAT_OBC_GPS_H
+
+#include <vector>
+
+#include "Buses/uart.hpp"
+
+namespace habsat::gps {
+// NVME type GGA message structure
+typedef struct {
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:14]: typedef not needed

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    char Mode;          // Mode  A= Autonomus D=DGPS, E=DR
+    char* checksum[5];  // Checksum of all chars between $ (start body message) and * (end body
+                        // message)
+
+} GPGGA_t;
+
+enum MessageType { RMC, VTG, GGA, GSA, GSV, GLL, None };
+
+typedef struct {
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:51]: typedef not needed

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    /*ignored because no information in documentation*/
+    // double  MagneticVariation;       //degrees E=east or W=west, East/West Indicator redundand
+
+    char Mode;          // Mode  A= Autonomus D=DGPS, E=DR
+    char* checksum[5];  // Checksum of all chars between $ (start body message) and * (end body
+                        // message)
+
+} GPGGA_t;
+
+enum MessageType { RMC, VTG, GGA, GSA, GSV, GLL, None };
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:49]: consider using enum class

Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:49]: dont use ALL_CAPS names for non macros

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+typedef struct {
+    std::string RMC;
+    std::string VTG;
+    std::string GGA;
+    std::string GSA;
+    std::string GSV;
+    std::string GLL;
+} GpsData;
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:58]: don't use ALL_CAPS for non macros

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+   public:
+    Gps(UART_HandleTypeDef* uart, DMA_HandleTypeDef* dma_usart_rx);
+    void init();
+    void callback();
+    GpsData getData();
+
+   private:
+    static const int rawBufferSize        = 250;
+    UART_HandleTypeDef* uart;
+    DMA_HandleTypeDef* dma_rx;
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:70]: prefer references to pointers

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+class Gps {
+   public:
+    Gps(UART_HandleTypeDef* uart, DMA_HandleTypeDef* dma_usart_rx);
+    void init();
+    void callback();
+    GpsData getData();
+
+   private:
+    static const int rawBufferSize        = 250;
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:68]: constexpr

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    std::string GSV;
+    std::string GLL;
+} GpsData;
+
+class Gps {
+   public:
+    Gps(UART_HandleTypeDef* uart, DMA_HandleTypeDef* dma_usart_rx);
+    void init();
+    void callback();
+    GpsData getData();
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:65]: wrong syntax
auto fn() -> GpsData
make sure to use provided .clang-tidy file

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    std::string VTG;
+    std::string GGA;
+    std::string GSA;
+    std::string GSV;
+    std::string GLL;
+} GpsData;
+
+class Gps {
+   public:
+    Gps(UART_HandleTypeDef* uart, DMA_HandleTypeDef* dma_usart_rx);
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:62]: prefer references

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    void init();
+    void callback();
+    GpsData getData();
+
+   private:
+    static const int rawBufferSize        = 250;
+    UART_HandleTypeDef* uart;
+    DMA_HandleTypeDef* dma_rx;
+    std::string rawMessageBuffor;
+    uint8_t rawBuffor[rawBufferSize];
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:72]: std::array

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    Gps(UART_HandleTypeDef* uart, DMA_HandleTypeDef* dma_usart_rx);
+    void init();
+    void callback();
+    GpsData getData();
+
+   private:
+    static const int rawBufferSize        = 250;
+    UART_HandleTypeDef* uart;
+    DMA_HandleTypeDef* dma_rx;
+    std::string rawMessageBuffor;
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:71]: don't use std::string. we do not allow dynamic allocations

File: Lib/Gps/Inc/Gps/gps.hpp [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    uint8_t rawBuffor[rawBufferSize];
+
+    GpsData GpsMessages;
+
+    void process();
+    MessageType matchMessageType(std::string& message);
+    GpsData retriveSingleMessage(std::string& rawMessage);
```
Comment (wiktorwieclaw) [inline Lib/Gps/Inc/Gps/gps.hpp:78]: don't use std::string

File: Lib/Gps/gps.cpp [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    if("VTG" == type)
+        return VTG;
+    if("GGA" == type)
+        return GGA;
+    if("GSA" == type)
+        return GSA;
+    if("GSV" == type)
+        return GSV;
+    if("GLL" == type)
+        return GLL;
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:85]: use switch

File: Lib/Gps/gps.cpp [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+    return None;
+}
+GpsData Gps::getData() { return GpsMessages; }
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:89]: wrong syntax style for functions

File: Lib/Gps/gps.cpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+        Terminal::pcTransmit(GpsMessages.VTG+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GGA+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GSA+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GSV+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GLL+"\r\n");
+
+    //Restart of transmition
+    HAL_UART_Receive_DMA(uart, reinterpret_cast<uint8_t*>(rawBuffor), rawBufferSize);
+}
+GpsData Gps::retriveSingleMessage(std::string & rawMessage) {
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:43]: wrong syntax style for functions

File: Lib/Gps/gps.cpp [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+    std::string m_RMC, m_VTG, m_GGA, m_GSA, m_GSV, m_GLL;
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:45]: this breaks naming convention

File: Lib/Gps/gps.cpp [hunk 22]
```diff
@@ -5,34 +5,87 @@
...
+    if("GGA" == type)
+        return GGA;
+    if("GSA" == type)
+        return GSA;
+    if("GSV" == type)
+        return GSV;
+    if("GLL" == type)
+        return GLL;
+
+    return None;
```
Comment (j-zygmunt) [inline Lib/Gps/gps.cpp:87]: you're checking same thing twice

File: Lib/Gps/gps.cpp [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace habsat::gps{
+
+Gps::Gps(UART_HandleTypeDef *uart, DMA_HandleTypeDef* dma_usart_rx) : uart(uart), dma_rx(dma_usart_rx){
+    init();
+}
+
+void gps::Gps::init() {
+    __HAL_UART_ENABLE_IT(uart, UART_IT_IDLE);
+    HAL_UART_Receive_DMA(uart, reinterpret_cast<uint8_t*>(rawBuffor), rawBufferSize);
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:18]: redundant cast (i think not sure)

File: Lib/Gps/gps.cpp [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+#include "Inc/Gps/gps.hpp"
+
+#include <terminal.hpp>
+#include <vector>
+
+namespace habsat::gps{
+
+Gps::Gps(UART_HandleTypeDef *uart, DMA_HandleTypeDef* dma_usart_rx) : uart(uart), dma_rx(dma_usart_rx){
+    init();
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:13]: don't do init in constructors. we will need to be able to control the initialization process later on

File: Lib/Gps/gps.cpp [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+void gps::Gps::callback() {
+
+    HAL_UART_DMAStop(uart);
+
+    uint8_t data_length  = rawBufferSize - __HAL_DMA_GET_COUNTER(dma_rx);
+
+    rawMessageBuffor = std::string(reinterpret_cast<char*>(rawBuffor));
+    rawMessageBuffor = rawMessageBuffor.substr(0,data_length);
+
+    GpsMessages = retriveSingleMessage(rawMessageBuffor);
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:29]: auto

File: Lib/Gps/gps.cpp [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+
+    //todo debug show messages
+    Terminal::pcTransmit("---  Gps Messages  ---\r\n");
+        Terminal::pcTransmit(GpsMessages.RMC+"\r\n");
+        Terminal::pcTransmit(GpsMessages.VTG+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GGA+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GSA+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GSV+"\r\n");
+        Terminal::pcTransmit(GpsMessages.GLL+"\r\n");
+
```
Comment (wiktorwieclaw) [inline Lib/Gps/gps.cpp:39]: don't use terminal in in /Lib/

**Review by wiktorwieclaw — COMMENTED** [review]
lot's of issues regarding correctness, style and readibility

**Pull request merged**